src/ngraph_assign_clusters.cc

```diff
@@ -83,23 +91,203 @@ namespace {
+  NGRAPH_VLOG(5) << "Contracting: " << src->name() << "[" << src->type_string()
+                 << " , " << edge->src_output() << "]@" << src_index << " -> "
+                 << dst->name() << "[" << dst->type_string() << " , "
+                 << edge->dst_input() << "]@" << dst_index;
+
+#if !defined(NGRAPH_TF_DISABLE_DEADNESS_CHECK)
+  string src_predicate = cluster_map[src]->predicate_string;
+  string dst_predicate = cluster_map[dst]->predicate_string;
+  NGRAPH_VLOG(5) << "Src pred: " << src_predicate
+                 << " ,Dst pred: " << dst_predicate;
+
+  std::string cluster_pred = GetMergedClusterPred(src_predicate, dst_predicate);
+
+  cluster_map[src]->predicate_string = cluster_pred;
+  // Update outgoing edges of the merged cluster
+  cluster_map[src]->outgoing_edges.insert(
+      cluster_map[dst]->outgoing_edges.begin(),
+      cluster_map[dst]->outgoing_edges.end());
+  cluster_map[src]->outgoing_edges.erase(edge);
+#endif
+
+  auto cluster_dst = cluster_map[dst];
+  // using cluster_map[dst]->nodes in the loop directly appears to
+  // invalidate the iterator when `node` == `dst`
+  // this happens with clang but not gcc
+  for (auto node : cluster_dst->nodes) {
+    cluster_map[src]->nodes.insert(node);
+    cluster_map[node] = cluster_map[src];
+  }
+}
+
+}  // namespace
+
+// Main Entry point for Cluster Assignment to the Node
+// Adds an attribute "_ngraph_cluster" (cluster_id) to each Node that can be
+// encapsulated
 Status AssignClusters(Graph* graph) {
   std::map<Node*, std::shared_ptr<Cluster>> cluster_map;
 
+#if !defined(NGRAPH_TF_DISABLE_DEADNESS_CHECK)
+  std::unique_ptr<DeadnessAnalysis> deadness_analyzer;
+  TF_RETURN_IF_ERROR(DeadnessAnalysis::Run(*graph, &deadness_analyzer));
+  // This map is used only for error checking
+  std::map<Node*, std::string> nodes_predicate_map;
+#endif
+
   GraphCycles gc;
 
-  for (auto node : graph->op_nodes()) {
+  // Initial Step: Each node is a cluster of its own
+  for (auto node : graph->nodes()) {
     int new_index = gc.NewNode();
-    NGRAPH_VLOG(5) << "Creating cycle graph node: " << new_index << " for "
-                   << node->name() << "[" << node->type_string() << "]";
     cluster_map[node] = std::make_shared<Cluster>();
     cluster_map[node]->index = new_index;
     cluster_map[node]->nodes.insert(node);
+    NGRAPH_VLOG(5) << "Creating graphcycle Node: " << new_index << " for "
+                   << node->name() << "[" << node->type_string() << "]";
+
+#if !defined(NGRAPH_TF_DISABLE_DEADNESS_CHECK)
+    // get predicate string for the node
+    string pred_string;
+    TF_RETURN_IF_ERROR(deadness_analyzer->GetNodePredicate(*node, pred_string));
+    nodes_predicate_map[node] = pred_string;
+    cluster_map[node]->predicate_string = pred_string;
+
+    cluster_map[node]->outgoing_edges = std::set<const Edge*>(
+        node->out_edges().begin(), node->out_edges().end());
+    NGRAPH_VLOG(5) << node->name() << "[" << node->type_string() << "]"
```

Comment:
minor: well `out_edges` does return an `EdgeSet` which is a set, so we might use that instead of copying to a STL set. But no need to rewrite all the types and functions for that now.

Reply:
It would still do a copy to a new EdgeSet that we use to define a cluster. May be i am missing the point here.